src/solver/impls/cvode/cvode.cxx

```diff
@@ -288,6 +294,89 @@ int CvodeSolver::init(int nout, BoutReal tstep) {
     CVodeSetMaxOrd(cvode_mem, mxorder);
   }
 
+  const auto max_nonlinear_iterations = (*options)["max_nonlinear_iterations"]
+    .doc("Maximum number of nonlinear iterations allowed by CVODE before reducing "
+         "timestep. CVODE default (used if this option is negative) is 3.")
+    .withDefault(-1);
+  if (max_nonlinear_iterations > 0) {
+    CVodeSetMaxNonlinIters(cvode_mem, max_nonlinear_iterations);
+  }
+
+  const auto apply_positivity_constraints = (*options)["apply_positivity_constraints"]
+    .doc("Use CVODE function CVodeSetConstraints to constrain variables - the constraint "
+         "to be applied is set by the positivity_constraint option in the subsection for "
+         "each variable")
+    .withDefault(false);
+#if not (SUNDIALS_VERSION_MAJOR >= 3 and SUNDIALS_VERSION_MINOR >= 2)
+  if (apply_positivity_constraints) {
+    throw BoutException("The apply_positivity_constraints option is only available with "
+                        "SUNDIALS>=3.2.0");
+  }
+#else
+  if (apply_positivity_constraints) {
+    std::vector<BoutReal> f2d_constraints;
+    f2d_constraints.reserve(f2d.size());
+    std::transform(begin(f2d), end(f2d), std::back_inserter(f2d_constraints),
+                   [](const VarStr<Field2D>& f2) {
+                     auto f2_options = Options::root()[f2.name];
+                     const auto value = f2_options["positivity_constraint"]
+                                        .doc(fmt::format(
+                                             "Constraint to apply to {} if "
+                                             "solver:apply_positivity_constraint=true. "
+                                             "Possible values are: none (default), "
+                                             "positive, non_negative, negative, or "
+                                             "non_positive.", f2.name))
+                                        .withDefault(positivity_constraint::none);
+                     switch (value) {
+                       case positivity_constraint::none: return 0.0;
+                       case positivity_constraint::positive: return 2.0;
+                       case positivity_constraint::non_negative: return 1.0;
+                       case positivity_constraint::negative: return -2.0;
+                       case positivity_constraint::non_positive: return -1.0;
+                       default: throw BoutException("Incorrect value for "
+                                                    "positivity_constraint");
+                     }
+                   });
```

Comment:
Might be worth pulling this out into a helper function? `VarStr<T>` is private, so you might need to template it like this:

```cpp
template <class FieldType>
std::vector<BoutReal> create_constraints(const std::vector<FieldType>& fields) {
    std::vector<BoutReal> field_constraints;
    field_constraints.reserve(fields.size());
    std::transform(begin(fields), end(fields), std::back_inserter(field_constraints),
                   [](const FieldType& field) {
...
```

or else make it a lambda and use `auto` for the lambda arguments.

Reply:
Good plan, done.